test/alternator/test_cql_rbac.py

```diff
@@ -0,0 +1,939 @@
+    url = dynamodb.meta.client._endpoint.host
+    config = dynamodb.meta.client._client_config
+    verify = not url.startswith('https')
+    ret = boto3.client('dynamodbstreams', endpoint_url=url, verify=verify,
+        aws_access_key_id=role, aws_secret_access_key=key,
+        region_name='us-east-1', config=config)
+    try:
+        yield ret
+    finally:
+        ret.close()
+
+# A basic test for creating a new role. The ListTables operation is allowed
+# to any role, so it should work in the new role when given the right password
+# and fail with the wrong password.
+def test_new_role(dynamodb, cql):
+    with new_role(cql) as (role, key):
+        with new_dynamodb(dynamodb, role, key) as d:
+            # ListTables should not fail (we don't care what is the result)
+            d.meta.client.list_tables()
+        # Trying to use the wrong key for the new role should fail to perform
+        # any request. The new_dynamodb() function can't detect the error,
+        # it is detected when attempting to perform a request with it.
+        with new_dynamodb(dynamodb, role, 'wrongkey') as d:
+            with pytest.raises(ClientError, match='UnrecognizedClientException'):
+                d.meta.client.list_tables()
+
+# A role without "login" permissions cannot be used to authenticate requests.
+# Reproduces #19735.
+def test_login_false(dynamodb, cql):
+    with new_role(cql, login=False) as (role, key):
+        with new_dynamodb(dynamodb, role, key) as d:
+            with pytest.raises(ClientError, match='UnrecognizedClientException.*login=false'):
+                d.meta.client.list_tables()
+
+# Quote an identifier if it needs to be double-quoted in CQL. Quoting is
+# *not* needed if the identifier matches [a-z][a-z0-9_]*, otherwise it does.
+# double-quotes ('"') in the string are doubled.
+def maybe_quote(identifier):
+    if re.match('^[a-z][a-z0-9_]*$', identifier):
+        return identifier
+    return '"' + identifier.replace('"', '""') + '"'
+
+# Unfortunately, permissions are currently cached for a fixed amount of
+# time, permissions_validity_in_ms, and changing them with GRANT or REVOKE
+# might not have any affect until this timeout. So we need to artificial
+# sleep in temporary_grant() below, and for that we need to know what
+# permissions_validity_in_ms.
+# To make tests fast, it should be configured (e.g. test/cql-pytest/run.py)
+# to be as low as possible, but this test should handle any configured
+# value.
+@cache
+def permissions_validity_in_ms(cql):
+    return 0.001 * int(list(cql.execute("SELECT value FROM system.config WHERE name = 'permissions_validity_in_ms'"))[0][0])
+
+# Convenience context manager for temporarily GRANTing some permission and
+# then revoking it.
+@contextmanager
+def temporary_grant(cql, permission, resource, role):
+    role = maybe_quote(role)
+    cql.execute(f"GRANT {permission} ON {resource} TO {role}")
+    time.sleep(permissions_validity_in_ms(cql)*2)
```

Comment:
or maybe disable permissions cache for those kind of tests?

Reply:
not sure how it applies in alterantor tests but in general in test.py you can set different cluster config for suites

Reply:
Interesting, I set permissions_validity_in_ms in the run script (you're write that I can do it in test.py too, I didn't, but perhaps I should) to 5ms, but perhaps I should just set it to 0ms and maybe this will mean caching will be completely disabled?

I was worried that setting the caching to very low will hurt the overall performance of the tests, but I checked that bringing it down from 100ms (the previous default in the tests) to 5ms didn't slow the overall test run, but I can also test 0ms.

Reply:
In the new version I'll upload shortly I removed these sleeps, changing them to retry loops. The tests can handle any setting of permissions_validity_in_ms, be it in 100ms (the existing default) or 0, but will run faster with lower setting.

I set permissions_validity_in_ms=5ms in one of the patches in this series for test/alternator/run (but not test.py), to make this test file take 2 seconds instead of 7 seconds. The difference between 7 seconds and 2 seconds was important for me for test/alternator/run (which takes just 30 seconds),  but nobody cares about it in the context of test.py.